File: firstMissingPositive/firstMissingPositive.java
/**
 *  https://leetcode.com/problems/first-missing-positive/description/
 *
 *  Runs in O(n) time with O(1) space
 */

class Solution {
    public int firstMissingPositive(int[] nums) {
        // Do an in-place sort of the array.
        for ( int i = 0; i < nums.length; i++ ) {
            // While nums[i] is in the range of 1-> size && 
            // is not in the correct place already...
            while ( nums[i] > 0 && nums[i] <= nums.length && nums[i] != nums[nums[i]-1] ) {
                int temp = nums[i]; // hold the value
                nums[i] = nums[temp-1]; // Put the value we're swapping with in nums[i]
                nums[temp-1] = temp; // Put nums[i] into nums[temp-1]
            }
        }
        // Search the array for the first value that isn't nums[i] = i+1
        for ( int i = 0; i < nums.length; i++ ) {
            if ( nums[i] != i+1 ) {
                return i+1;
            }
        }
        return nums.length+1;
    }
}

File: firstMissingPositive/firstMissingPositive.cpp
/**
 *  https://leetcode.com/problems/first-missing-positive/description/
 *
 *  Runs in O(n) time with O(1) space
 */

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        // Iterate through the vector
        for ( int i = 0; i < nums.size(); i++){
            // Swap the value of the current index to it's correct position
            // or if the current position contains an invalid number (out of range)
            // The ordering assumes that nums[0] should contain 1, nums[1] = 2 .. etc
            while (nums[i] > 0 && nums[i] <= nums.size() && nums[i] != nums[nums[i]-1] ) {
                // Do swap while checking for repeated numbers in the array.
                int temp = nums[i];
                nums[i] = nums[temp - 1]; // take nums[i]-1 to swap with
                nums[temp - 1] = temp;
            }
        }
        // Each index should contain the correct number
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] != i+1){
                // The first index that contains the incorrect number is the answer.
                return i+1;
            }
        }
        // If everything is correct in the vector, then the answer is size + 1
        return nums.size() + 1;
    }
};
